common/utils.go

```diff
@@ -58,40 +60,108 @@ func FindNetDevs(processID int, match func(string) bool) ([]NetDev, error) {
 	defer ns.Close()
 
 	err = WithNetNS(ns, func() error {
-		links, err := netlink.LinkList()
-		if err != nil {
-			return err
-		}
-		for _, link := range links {
-			if match(link.Attrs().Name) {
-				addrs, err := netlink.AddrList(link, netlink.FAMILY_V4)
+		return forEachLink(func(link netlink.Link) error {
+			if match(link) {
+				netDev, err := linkToNetDev(link)
 				if err != nil {
 					return err
 				}
-
-				netDev := NetDev{MAC: link.Attrs().HardwareAddr}
-				for _, addr := range addrs {
-					netDev.CIDRs = append(netDev.CIDRs, addr.IPNet)
-				}
-				netDevs = append(netDevs, netDev)
+				netDevs = append(netDevs, *netDev)
 			}
-		}
-		return nil
+			return nil
+		})
 	})
 
 	return netDevs, err
 }
 
+func forEachLink(f func(netlink.Link) error) error {
+	links, err := netlink.LinkList()
+	if err != nil {
+		return err
+	}
+	for _, link := range links {
+		if err := f(link); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func linkToNetDev(link netlink.Link) (*NetDev, error) {
+	addrs, err := netlink.AddrList(link, netlink.FAMILY_V4)
+	if err != nil {
+		return nil, err
+	}
+
+	netDev := &NetDev{MAC: link.Attrs().HardwareAddr}
+	for _, addr := range addrs {
+		netDev.CIDRs = append(netDev.CIDRs, addr.IPNet)
+	}
+	return netDev, nil
+}
+
 // Lookup the weave interface of a container
 func GetWeaveNetDevs(processID int) ([]NetDev, error) {
-	return FindNetDevs(processID, func(name string) bool {
-		return strings.HasPrefix(name, "ethwe")
+	// Bail out if this container is running in the root namespace
+	nsToplevel, _ := netns.GetFromPid(1)
+	nsContainr, _ := netns.GetFromPid(processID)
+	if nsToplevel.Equal(nsContainr) {
+		return nil, nil
+	}
+
+	var weaveBridge netlink.Link
+	forEachLink(func(link netlink.Link) error {
+		if link.Attrs().Name == "weave" {
+			weaveBridge = link
+		}
+		return nil
+	})
+	if weaveBridge == nil {
+		return nil, fmt.Errorf("Cannot find weave bridge")
+	}
+	// Scan devices in root namespace to find those attached to weave bridge
+	indexes := make(map[int]struct{})
+	err := forEachLink(func(link netlink.Link) error {
+		if link.Attrs().MasterIndex == weaveBridge.Attrs().Index {
+			indexes[link.Attrs().Index] = struct{}{}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return FindNetDevs(processID, func(link netlink.Link) bool {
+		// For checking, rely on index number of veth end inside namespace being -1 of end outside
```

Comment:
No, I don't think it is very safe.  While veth ifindexes will tend to be consecutive, I don't think there is any guarantte about which comes first.  And the might not be consecutive.  Consider the case where two veth pairs are created concurrently.  Also, it looks like it might be possible to specify a ifindex when creating a new dev.  Also, ifindex allocation can wrap.

Netlink provides a way to find out the peer of a veth.  This wasn't always there, but IIRC it's been around for a while. We should at least try that method.